id24: SBARDEF

rfomin opened the pull request:
Legacy of Rust fullscreen HUD:

![woof0002](https://github.com/user-attachments/assets/e983d82d-c49a-41db-869d-902f64693d63)

This is very incomplete, but I have a few suggestions/questions:
1. Should we remove Crispy HUD?
2. We can remove the Boom HUD and WOOFHUD, but move the new useful widgets to SBARDEF. This will be a non-standard extension.

Note that `SBARDEF` also defines the classic status bar. The `SBARDEF` in id1.wad consists of 1955 lines, it is difficult to edit by hand.

fabiangreffrath commented:
Can we define multiple different HUDs with SBARDEF? Can we use it to define both the classical status bar and the Crispy HUD and switch between both?

fabiangreffrath commented:
How does it distinguish between 4:3 and widescreen alignment?

rfomin commented:
> Can we define multiple different HUDs with SBARDEF? Can we use it to define both the classical status bar and the Crispy HUD and switch between both?

Currently in LoR defined 3 statusbar HUD's, `statusbars[0]` is replication of classic statusbar, `statusbars[1]` is fullscreen HUD, similar to Crispy one (that's why I think we don't need Crispy variant), `statusbars[2]` is empty. This is already working in this PR.

> How does it distinguish between 4:3 and widescreen alignment?

It doesn't, but there is support for "tall" screens with aspect ratio < 4:3 ("HUD mode" option, `conditions` 18). We can add a widescreen mode option or just drop this functionality.

MrAlaux commented on src/st_stuff.c:
```diff
@@ -318,10 +322,687 @@ int             st_keyorskull[3];
 // STATUS BAR CODE
 //
 
+static void LoadFacePathes(void)
```
Should be `LoadFacePatches`.

rfomin replied:
Fixed in https://github.com/fabiangreffrath/woof/pull/1916/commits/c40caecd5428fa13fd15c884b0a4be3e2b083e76

fabiangreffrath commented:
> Currently in LoR defined 3 statusbar HUD's, `statusbars[0]` is replication of classic statusbar, `statusbars[1]` is fullscreen HUD, similar to Crispy one (that's why I think we don't need Crispy variant), `statusbars[2]` is empty. This is already working in this PR.

I assume it isn't too hard to extend this to more than 3 HUDs, e.g. Fullscreen HUD with and without face widget?

> It doesn't, but there is support for "tall" screens with aspect ratio < 4:3 ("HUD mode" option, `conditions` 18). We can add a widescreen mode option or just drop this functionality.

Alright. I wonder why this hasn't been considered by the original authors yet.

rfomin commented:
> I assume it isn't too hard to extend this to more than 3 HUDs, e.g. Fullscreen HUD with and without face widget?

Yes, but why? I like Doom guy face 😄 

> Alright. I wonder why this hasn't been considered by the original authors yet.

It doesn't look good on a very wide monitor. I think a "virtual" 320x200 resolution for the HUD makes sense.

fabiangreffrath commented on src/st_sbardef.c:
```diff
@@ -0,0 +1,332 @@
+    out->type = JS_GetInteger(type);
+
+    char lump[9] = {0};
+    boolean found = false;
```
This needs to be an `int`.

rfomin replied:
Fixed in https://github.com/fabiangreffrath/woof/pull/1916/commits/77ae8d3df3be9c2d97b78eaef09ff7cdf4ba7af3

fabiangreffrath commented:
Automap is broken.

rfomin commented:
> Automap is broken.

Mostly works in latest commit.

> How do I switch between HUDs?

Press "+/-"

fabiangreffrath commented:
It crashes with `-file ~/id24res.wad`.

fabiangreffrath commented:
> Mostly works in latest commit.

![woof0001](https://github.com/user-attachments/assets/5a5c038b-35b0-4081-ab89-e6c49e3ee5e7)

fabiangreffrath commented:
> Press "+/-"

Could we please map this to F5?

rfomin commented:
> > Press "+/-"
> 
> Could we please map this to F5?

This is from the spec, the full screen HUD is tied to the screen size. Sure, we can add it later.

Did I mention that this PR is very incomplete? 😆

rfomin commented:
> It crashes with `-file ~/id24res.wad`.

This is a bug in id24res.wad, it has a `"param": -1879048192,` line in it. I'm not sure if we should check this.

fabiangreffrath commented:
In the weapon box on the status bar, all weapons always appear as available (i.e. yellow instead of gray).

rfomin commented:
> In the weapon box on the status bar, all weapons always appear as available (i.e. yellow instead of gray).

Fixed in https://github.com/fabiangreffrath/woof/pull/1916/commits/8aaced8acb8bb49b25c93298032d97bfa59f220a

rfomin commented on base/all-all/sbardef.lmp:
```diff
@@ -123,7 +301,117 @@
+                  "condition": 22,
+                  "param": 1
+                }
+              ],
```
@fabiangreffrath I implement Boom colors by extending SBARDEF. We will have to increase the lump version, and it will not work with the default Legacy of Rust WAD. So users who want this feature will have to overwite SBARDEF.

fabiangreffrath replied:
> I implement Boom colors by extending SBARDEF.

Thank you! Doesn't seem to work with my build, though. 



> and it will not work with the default Legacy of Rust WAD. So users who want this feature will have to overwite SBARDEF.

I find that reasonable, since Legacy of Rust decided to ship its own HUD definitions, and we ship ours. 

> We will have to increase the lump version

Should we introduce a Woof namespace probably?

rfomin replied:
> > I implement Boom colors by extending SBARDEF.
> 
> Thank you! Doesn't seem to work with my build, though.

Is your woof.pk3 updated (it should contain sbardef.lmp). Did you enable “Colored Numbers” in the menu?

fabiangreffrath replied:
I think so:

https://github.com/user-attachments/assets/56a8a65a-0c63-4393-a8e7-a4678fdab612

rfomin replied:
> I think so:

But it works, so far I've only done it for the classic status bar 😄

What do you think about this approach? I want to remove the color configuration options (except `sts_colored_numbers`), users can edit SBARDEF themselves. But editing it is not very fun.

fabiangreffrath replied:
> But it works, so far I've only done it for the classic status bar 😄

Ah, this explains a lot! :grin: 

> What do you think about this approach? I want to remove the color configuration options (except `sts_colored_numbers`), users can edit SBARDEF themselves. But editing it is not very fun.

I think it's reasonable. I don't think I have ever edited the threshold values in my entire life, but I'm not sure if others ever did.

However, we still need the health-to-color logic for the health-colored crosshair, so maybe it's still better to leave this in a function?

Another question: If we have, say, 200 health, do we draw the health widget four times? I mean, once in red, because 200 >= 0, once in yellow, because 200 >= 25, once in green, because 200 >= 50 and finally in blue, because 200 >= 100?

fabiangreffrath replied:
Also, health and armor "turn to stone" (i.e. get drawn in gray) in invincibility mode. We shouldn't introduce too many regressions for people who use the colored status bar feature.

rfomin replied:
> However, we still need the health-to-color logic for the health-colored crosshair, so maybe it's still better to leave this in a function?

Sure, we'll keep this function.

> Another question: If we have, say, 200 health, do we draw the health widget four times? I mean, once in red, because 200 >= 0, once in yellow, because 200 >= 25, once in green, because 200 >= 50 and finally in blue, because 200 >= 100?

Yes 😆 I need to add additional conditions (`>=` and `<=`).

fabiangreffrath commented on base/all-all/sbardef.lmp:
```diff
@@ -123,7 +301,117 @@
+              "alignment": 2,
+              "font": "BigRed",
+              "tranmap": null,
+              "translation": "CRBLUE",
```
We use `CRBLUE2` for the BigRed font, please.

rfomin replied:
Fixed in 5e954e2

fabiangreffrath commented on src/st_stuff.c:
```diff
@@ -499,6 +502,31 @@ static boolean CheckConditions(sbarcondition_t *conditions, player_t *player)
+            case sbc_ammogreaterequal_percent:
+                {
+                    ammotype_t type = weaponinfo[player->readyweapon].ammo;
+                    if (type != am_noammo)
```
This also needs a check for `player->maxammo[type] > 0`. Also, I am missing the ability to turn ammo blue if it is more than the non-backpack max-value.

rfomin replied:
Fixed in current version.

fabiangreffrath commented on src/st_stuff.c:
```diff
@@ -499,6 +502,31 @@ static boolean CheckConditions(sbarcondition_t *conditions, player_t *player)
+                break;
+
+            case sbc_armorgreaterequal:
+                result &= player->armorpoints >= cond->param;
```
I'd prefer color by armor type here.

fabiangreffrath replied:
If this is getting too complicated, we could still change the `translation` value by function, right? Is it possible to define some kind of "callback function" for the status bar elements that gets called whenever they are rendered?

rfomin replied:
I agree, fixed here: https://github.com/fabiangreffrath/woof/pull/1916/commits/5e954e257fa5dbc8dfc9ff63aa47264a4475d40a

fabiangreffrath commented on base/all-all/sbardef.lmp:
```diff
@@ -90,6 +90,100 @@
+                  "param": 0
+                },
+                {
+                  "condition": 20,
```
What I meant with "namespace": Is it possible to e.g. introduce a "condition-woof" field for conditions >= 20, that other ports simply don't parse? Or will they error out when they encounter an unknown field name?

rfomin replied:
We can make a large container (“canvas”) with a “boom” condition, or simply define `statusbar[3]`, which is ignored in the current specification.

Alternatively, we can increase the lump version so that the KEX port ignores our SBARDEF.

fabiangreffrath commented:
> > > Press "+/-"

I'd prefer to use +/- to strictly increase/decrease HUDs and F5 to toggle through them, pretty much as it already is.

rfomin commented:
Boom coloring and most of the old stbar features should work now. I didn't implement blinking keys and solid background.

I started to move HUD widgets to SBARDEF, the goal is to delete hu_* files completely. Athought this PR became big (most of it is giant `base/all-all/sbardef.lmp` file), so I'm looking for feedback. Should I continue with this PR or maybe split it into parts?

fabiangreffrath commented on src/st_stuff.c:
```diff
@@ -271,1207 +121,1487 @@ int armor_green;   // armor amount above is blue, below is green
+                    return;
+                }
+
+                int maxammo = player->maxammo[type];
```
Please treat `player->maxammo[type] == 0` the same as `type == am_noammo`.

rfomin replied:
Fixed in https://github.com/fabiangreffrath/woof/pull/1916/commits/692c436a252ca74ad669ffb0f5fd27e13e222711

fabiangreffrath commented on base/all-all/dig033.png:
Why? What's wrong with our hud font?

rfomin replied:
We have an incompatible naming scheme for it compared to DSDA-Doom/PrBoom+ or IWAD fonts. I think the big redesign of the HUD is an opportunity to bring it in line with other ports. See the fonts section in SBARDEF:
```json
"numberfonts":
    [
      {
        "name": "BigRed",
        "type": 0,
        "stem": "STT"
      },
      {
        "name": "SmallGrey",
        "type": 0,
        "stem": "STG"
      },
      {
        "name": "SmallYellow",
        "type": 0,
        "stem": "STYS"
      }
    ],
    "hudfonts":
    [
      {
        "name": "ConFont",
        "type": 2,
        "stem": "STCFN"
      },
      {
        "name": "SmallFont",
        "type": 0,
        "stem": "DIG"
      }
    ],
```

fabiangreffrath commented:
> Should I continue with this PR or maybe split it into parts?

Not sure where to draw the line when everything is connected with each other. Maybe the best place for a split is between "adding new functionality" and "removing old functionality". 🤷

fabiangreffrath commented on src/st_stuff.c:
```diff
@@ -861,19 +836,113 @@ static void UpdateAnimation(sbarelem_t *elem)
+                ammotype_t type = weaponinfo[player->readyweapon].ammo;
+                if (type == am_noammo)
+                {
+                    return;
```
Will `cr` still be uninitiaized at this point?

fabiangreffrath replied:
Never mind, I misread the function that it would return the value of `cr`.

rfomin commented:
> Not sure where to draw the line when everything is connected with each other. Maybe the best place for a split is between "adding new functionality" and "removing old functionality". 🤷

Honestly, I want to do the whole thing in one PR, I just think it might be hard to review. Indeed, the `ST_*` and `HU_*` code does the same thing, in my opinion.

fabiangreffrath commented on base/all-all/sbardef.lmp:
```diff
@@ -866,6 +866,21 @@
               "conditions": null,
               "children": null
             }
+          },
```
Does this mean that whenever a PWAD provides its own SBARDEF lump, such as Legacy of Rust, we lose the ability to show messages?

rfomin replied:
Right, I forgot about that! Fixed here: https://github.com/fabiangreffrath/woof/pull/1916/commits/ede3bc063a52e3cfec31ddc99f19ba205092bfbf

rfomin commented:
It's almost done.

What is missing:
* `hud_widget_font` option. The font can be set for any widget in `SBARDEF`, however we can force `hu_font` for some widgets.
* `hud_blink_keys` I haven't figured out how it works.
* `hud_widget_layout` vertical layout. Not sure if we need this.
* `message_centered` The message widget can be anywhere, but we can reset the position and force it to be centred.
* `hud_backpack_thresholds` enabled by default.

fabiangreffrath commented on base/all-all/woofhud.lmp:
Could you also remove `WOOFHUD` documentation in `docs/woofhud.md` and its mention in `README.md`?

rfomin replied:
Done in https://github.com/fabiangreffrath/woof/pull/1916/commits/ec9ae274b7fdc8cc94956b4ed6616a9757a89e40

fabiangreffrath commented on base/all-all/sbardef.lmp:
Could you add copyright/license information for the new lumps to `README.md`?

rfomin replied:
The original `SBARDEF` has no copyright/author information, so I don't know what to write in the `Copyright:` line. It's just a configuration file, is it this necessary?

ID24 spec/resources licensed under CC0.

fabiangreffrath replied:
> The original `SBARDEF` has no copyright/author information, so I don't know what to write in the `Copyright:` line. It's just a configuration file, is it necessary?

Well, we cannot just take what we've found somewhere on the net.

> ID24 spec/resources licensed under CC0.

At least that's a license, but there must be someone who grants that license.

fabiangreffrath replied:
Google Drive states that gooberman is the author, so this would be the obvious choice.

fabiangreffrath commented:
I am missing the ability to render fullscreen but show no HUD at all.

fabiangreffrath commented:
The "level announcement" should reuse the Automap level title widget.

rfomin commented:
> I am missing the ability to render fullscreen but show no HUD at all.

Should it only trigger with F5 or should I also add the 11th `screenblocks`?

> The "level announcement" should reuse the Automap level title widget.

This is how it works in DSDA-Doom, and I like it. It's also much easier to implement 😄

![doom00](https://github.com/user-attachments/assets/6fce3b98-242c-41ab-b4be-1972ef5b221b)

fabiangreffrath commented:
> Should it only trigger with F5 or should I also add the 11th `screenblocks`?

Yes, please, just as it was before.

> This is how it works in DSDA-Doom, and I like it. It's also much easier to implement 😄

But then it switches back and forth if the Automap is opened while it is displayed.

rfomin commented:
> All messages are grey.

Fixed in https://github.com/fabiangreffrath/woof/pull/1916/commits/1f9cb2675110a6b021ee88bc780d0cf4bfdbc241

> I am missing the ability to render fullscreen but show no HUD at all.

Fixed in https://github.com/fabiangreffrath/woof/pull/1916/commits/bc646627682b82310d4a6d4f77bcfbc8eae4d819

> > This is how it works in DSDA-Doom, and I like it. It's also much easier to implement 😄
> 
> But then it prints two lines if the Automap is opened while it is displayed.

To be honest, I don't think it's a problem. Can we leave it as it is, please?

fabiangreffrath commented:
Alright, now how do I e.g. move the monsec widget to the top left corner?

rfomin commented:
> Alright, now how do I e.g. move the monsec widget to the top left corner?

Set `x = 0`, `y = 8` (because we don't want to overwrite messages widget), `aligment = 16` (`sbe_wide_left` flag):

```json
{
  "widget":
  {
    "x": 0,
    "y": 8,
    "alignment": 0,
    "tranmap": null,
    "translation": null,
    "type": 0,
    "font": "Digits",
    "conditions": null,
    "children": null
  }
}
```
To center: `x = 160`, `y = 0`, `aligment = 1` (`sbe_h_middle` flag).

There is no automatic widget movement like in WOOFHUD except for the widescreen flags (`sbe_wide_left`, `sbe_wide_right`) I added. We can emulate this by adding more conditions, as I did with the map title widget.

fabiangreffrath commented on base/all-all/sbhuddef.lmp:
```diff
@@ -0,0 +1,257 @@
+{
```
What exactly is the difference between `sbhuddef` and `sbardef` again?

rfomin replied:
There are no HUD widgets in `sbardef` v1.0.0 (Legacy of Rust), so we will have to load them somehow. `sbhuddef` is only HUD widgets, previously it was a table in the code.

fabiangreffrath replied:
Alright, thanks!

fabiangreffrath commented on base/all-all/sbardef.lmp:
```diff
@@ -0,0 +1,1974 @@
+    "statusbars":
+    [
+      {
+        "height": 200,
```
Shoudn't this be `200 - 32` for the status bar hud?

rfomin replied:
It's just `"height": 32` in v1.0.0 lumps. I've  I made it 200 since we have HUD widgets.

fabiangreffrath replied:
Okay, I see. It's complicated...

fabiangreffrath commented on src/st_stuff.c:
```diff
@@ -86,1400 +84,1712 @@
+            int ch = *str++;
 
-  st_oldhealth = -1;
+            if (ch == '\x1b' && str)
```
`&& *str` ?

rfomin replied:
Fixed in https://github.com/fabiangreffrath/woof/pull/1916/commits/deb884d89cfc874bfb4bd09bdad259575cf06a14, thanks! Perhaps one day we should consolidate all the string drawing functions.

fabiangreffrath replied:
There is another occurrence in line 758.

rfomin replied:
Fixed here: https://github.com/fabiangreffrath/woof/pull/1916/commits/78488d25687334499b960158b92d416e495cb204

fabiangreffrath commented on src/st_stuff.c:
```diff
@@ -86,1400 +84,1712 @@
-// keycard widgets
-static st_multicon_t w_keyboxes[3];
+    // [FG] support face gib animations as in the 3DO/Jaguar/PSX ports
+    int painface;
```
Let's have just one variable of that name in the function context.

rfomin replied:
FIxed in https://github.com/fabiangreffrath/woof/pull/1916/commits/6f49928acaec7ad5661e729e557502b2977f0d82

fabiangreffrath commented:
The total level time is wrong on the intermission screen (there was a comment about this in the original code):

![woof0001](https://github.com/user-attachments/assets/35d5f4d2-8641-436b-9f94-3fa5c06a0128)

Also, the level time widget is hovering one line above the status bar if the level stats widget is not enabled at the same time. Is there anything we can do about it?

rfomin commented on src/st_stuff.c:
```diff
@@ -86,1400 +84,1712 @@
-    static int odelta = 0;
+    if (!st_refresh_background)
+    {
+        V_CopyRect(0, 0, st_backing_screen, video.unscaledw, ST_HEIGHT, 0,
```
Fixed here: https://github.com/fabiangreffrath/woof/pull/1916/commits/78488d25687334499b960158b92d416e495cb204

rfomin commented:
> The total level time is wrong on the intermission screen (there was a comment about this in the original code)

Fixed here: https://github.com/fabiangreffrath/woof/pull/1916/commits/34c497f8fdb2a2771fde318c210511c7ba6b1a61

> Also, the level time widget is hovering one line above the status bar if the level stats widget is not enabled at the same time. Is there anything we can do about it?

Fixed here: https://github.com/fabiangreffrath/woof/pull/1916/commits/b0398157d9b97dea8bb1e8c882b7db022f1a73dc

rfomin commented on base/all-all/sbardef.lmp:
```diff
@@ -874,6 +874,34 @@
+            }
+          },
+          {
+            "widget":
```
> but this isn't going to scale well.

Why not, it's a very small tree compared to typical Doom levels for example. I doubt we'll have many of these conditions/copies. Yes, it can't effectively replace the functionality of WOOFHUD, but I think it's good enough.

fabiangreffrath replied:
I think it's too much hard-coded to the default layout. For example, if I decide to move the level stats to the upper left, but leave the level time in the bottom left, will this work with that approach?

NightFright2k19 commented:
Maybe it would make sense to make a poll in the forums at some point about whether people are OK with WOOFHUD support to be removed. 

Personally I'm using it, even though I probably wouldn't whine too much if it really needs to go. I realize that having too many statusbar systems in the port is kinda overkill and bloats the code needslessly. I've taken a first look at SBARDEF and its structure. Honestly it's complicated AF and as you wrote already, it's rather unlikely it's going to become a popular method of statusbar modding unless there's a way to simplify it for the average user. WOOFHUD is much cleaner and straight-forward, even though I guess people rather stick to the original statusbar. (I have to admit I reverted back to it by myself, however I will try the Nightdive-style fullscreen display once it's available.)

rfomin commented:
> At first glance, SBARDEF looks complicated AF and as you wrote already, it's rather unlikely to turn into a popular method of statusbar modding unless it becomes more accessible.

It should not be edited by hand. I will make a visual editor for SBARDEF:

https://github.com/user-attachments/assets/b8b10480-858c-484b-8d73-6c556bc9d574

fabiangreffrath commented:
> Maybe it would make sense to make a poll in the forums at some point about whether people are OK with WOOFHUD support to be removed. 

Nope, sorry. This won't be decided by a poll. This will be decided by the devs who write and maintain that code in their free time.

NightFright2k19 commented:
It is sad that such a thing is apparently necessary and one wonders what was the thought behind the concept in the first place regarding modding capabilities. 

That bring said, I highly appreciate this effort. Should testing for this tool should be a thing, I'd sign up for it. Kinda curious what is actually possible with this new format. (If I understood correctly, the SBARDEF in id24res.wad contains a modified statusbar with support for the new LoR weapons and ammo enabled, so it's not a valid template for the vanilla Doom statusbar.)

rfomin commented:
> Fair enough. After all, I assume the Boom fullscreen HUD isn't going to be removed entirely, just the means to edit it? Otherwise, I hope this new format will at least allow to faithfully recreate it.

We keep all the new “widgets” like the level stat string, timer, etc. In some capacity it will be possible to recreate the old Boom HUD, I will try it when the editor is finished.

> Speaking of the editor now: It is sad that such a thing is apparently necessary and one wonders what was the thought behind the concept in the first place regarding modding capabilities.

I assume the id24 spec is inspired by modern game development that tends not to use scripting languages (e.g. Unreal Engine blueprints, etc.).

fabiangreffrath commented:
The widgets are missing the gap to the status bar and the left screen edge. This is how it looks in current master:

![Bildschirmfoto vom 2024-10-09 19-00-48](https://github.com/user-attachments/assets/1c51f11a-7432-4108-8fb6-af1fffa27a88)

And this is how it looks on this branch:

![Bildschirmfoto vom 2024-10-09 19-02-38](https://github.com/user-attachments/assets/b14de534-1fd8-48a5-b9cc-9592401964a6)

rfomin commented:
> The widgets are missing the gap to the status bar and the left screen edge.

Fixed in https://github.com/fabiangreffrath/woof/pull/1916/commits/28b1fbf5b9601cc3725191eaf51b39f2613148a7

fabiangreffrath commented on src/st_sbardef.c:
```diff
@@ -0,0 +1,511 @@
+        return out;
+    }
+
+    json = JS_Open("SBHUDDEF", "hud", (version_t){1, 0, 0});
```
If we allow to load SBARDEF lumps with version < 1.1.0 (i.e. without HUD widget definitions) and extend them by loading our own HUD widget definitions here, then why do we have to extend our own SBARDEF lump with HUD widget definitions and increase its version to 1.1.0 in the first place?

rfomin replied:
Yes, we can move all widgets to our own lump. But it's more convenient to have them integrated in SBARDEF, for example the user can have different `canvas` elements with different conditions, containing different STBAR and HUD widgets.

We also support Boom translation lumps and additional widescreen alignment, so we should increase the version anyway.

NightFright2k19 commented:
Something that just came to my mind:
One of the current shortcomings of the Nightdive port is that the fullscreen HUD still is only optimized for 4:3. Would the upcoming editor allow elements to be moved to widescreen areas as well?

fabiangreffrath commented:
> > To be honest, I don't think it's a problem. Can we leave it as it is, please?
> 
> He, okay. 😉

Maybe the map announcement message could be made the same color as the map title widget on the automap?

fabiangreffrath commented on base/all-all/sbardef.lmp:
```diff
@@ -930,7 +930,7 @@
             "widget":
             {
-              "x": 0,
+              "x": 2,
```
Would it be possible to stringify entries like e.g. `"alignment": 16` somehow? I think this would raise the acceptance of the new lump syntax. 

Something like `"alignment": "left-align"` or similar?

rfomin replied:
Even with these changes, the JSON format will be difficult to edit by hand. And it is not designed for that.

The editor is working, I'm just busy and took a break for a week.

fabiangreffrath commented:
On this branch the Automap title line does not respect the Automap color scheme.

fabiangreffrath commented:
> On this branch the Automap title line does not respect the Automap color scheme.

This fixes it for me (and #1936 while at it):

```patch
diff --git a/src/am_map.c b/src/am_map.c
index 9bd8afdb..67b5c80b 100644
--- a/src/am_map.c
+++ b/src/am_map.c
@@ -2387,6 +2387,8 @@ void AM_ColorPreset(void)
     mapcolor_secr = I_GetNearestColor(playpal, 255, 0, 255);
     mapcolor_revsecr = I_GetNearestColor(playpal, 119, 255, 111);
   }
+
+  ST_ResetTitle();
 }
 
 void AM_BindAutomapVariables(void)
diff --git a/src/st_widgets.c b/src/st_widgets.c
index 83bbe4cf..1f90f016 100644
--- a/src/st_widgets.c
+++ b/src/st_widgets.c
@@ -553,7 +553,9 @@ static char title_string[HU_MAXLINELENGTH];
 
 void ST_ResetTitle(void)
 {
-    title_string[0] = '\0';
+    title_string[0] = '\x1b';
+    title_string[1] = '0' + hudcolor_titl;
+    title_string[2] = '\0';
 
     char *s;
 
@@ -615,7 +617,10 @@ void ST_ResetTitle(void)
 
     if (hud_map_announce && leveltime == 0)
     {
-        displaymsg("%s", title_string);
+        if (gamemapinfo && U_CheckField(gamemapinfo->author))
+            displaymsg("%s by %s", title_string, gamemapinfo->author);
+        else
+            displaymsg("%s", title_string);
     }
 }
 
```

rfomin commented:
> This fixes it for me (and https://github.com/fabiangreffrath/woof/issues/1936 while at it):

Done in https://github.com/fabiangreffrath/woof/pull/1916/commits/01a158d3acda2307e6d5c14380a31bd93e038872, thanks!

> Maybe the map announcement message could be made the same color as the map title widget on the automap?

Less gold the better 😄

NightFright2k19 commented:
If I may ask: What's the status of the editor? It's ready by now?

rfomin commented:
> If I may ask: What's the status of the editor? It's ready by now?

It's working, but it's not ready for release.

NightFright2k19 commented:
Well, since it's basically needless before the feature is implemented, that's OK. And I doubt anyone would like to experiment with SBARDEF in the Nightdive port right now.

fabiangreffrath reviewed (approved):
🤞

fabiangreffrath commented on src/CMakeLists.txt:
```diff
@@ -282,11 +281,13 @@ if(MSVC)
+    "${CMAKE_CURRENT_BINARY_DIR}/woof.pk3" --format=zip .
     WORKING_DIRECTORY "${CMAKE_SOURCE_DIR}/base")
 
+add_dependencies(woof woof-base)
```
This *always* rebuilds the PK3 file:
```
$ git branch | grep ^\*
* id24_sbardef

$ ninja -C build/
ninja: Entering directory `build/'
[1/1] cd /home/fabian/Source/woof/base && /usr/bin/cmake -E tar cf /home/fabian/Source/woof/build/src/woof.pk3 --format=zip .

$ ninja -C build/
ninja: Entering directory `build/'
[1/1] cd /home/fabian/Source/woof/base && /usr/bin/cmake -E tar cf /home/fabian/Source/woof/build/src/woof.pk3 --format=zip .
```

rfomin replied:
Fixed in https://github.com/fabiangreffrath/woof/pull/1916/commits/0f1f8d921720b2e48eb63e7a6269eac0adb91921

fabiangreffrath replied:
Thank you!
